1990 New Jersey Congressional Districts

brianlzhou opened the pull request:
# 1990 New Jersey Congressional Districts

## Redistricting requirements
In New Jersey, we consult [NCSL Redistricting Law 2000](https://web.archive.org/web/20041216185957/https://www.senate.mn/departments/scr/redist/red2000/Tab5appx.htm) and impose the following constraints. In our simulations, districts must:

1. be contiguous
1. have equal populations
1. be geographically compact
1. preserve county and municipality boundaries as much as possible


### Algorithmic Constraints
We enforce a maximum population deviation of 0.5%.

## Data Sources
Data for New Jersey comes from the [ALARM Project's update](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/ZV5KF3) to [The Record of American Democracy](https://road.hmdc.harvard.edu/).

## Pre-processing Notes
No manual pre-processing decisions were necessary.

## Simulation Notes
We sample 50,000 districting plans for New Jersey across 5 independent runs of the SMC algorithm.
We then thinned the number of samples to 5,000. 
No special techniques were needed to produce the sample.

## Validation

<img width="3000" height="5400" alt="image" src="https://github.com/user-attachments/assets/6f452ab5-a7cc-4581-a17b-dfe9a5cda325" />

<img width="881" height="533" alt="image" src="https://github.com/user-attachments/assets/7e7e00c4-3356-4038-922c-6757407f2be2" />


```
SMC with Merge-Split MCMC Steps: 5,000 sampled plans of 13 districts on 1,944 units
Plans sampled on Spanning Forest Space using the Uniform Valid Edge forward kernel.
SMC Parameters:
• `weight_type` = optimal
• `seq_alpha` = 1
• `pop_temper` = 0.01
Mergesplit Parameters:
• `total_ms_steps` = 5
• `mh_accept_per_smc` = 50
• `pair_rule` = uniform
Plan diversity 80% range: 0.34 to 0.57
Largest R-hat values for ordered summary statistics:
    comp_bbox_reock           comp_edge         comp_polsby       county_splits         muni_splits 
              1.002               1.001               1.002               1.001               1.001 
            ndshare                 ndv                 nrv            plan_dev            pop_aian 
              1.011               1.002               1.004               1.000               1.009 
          pop_asian           pop_black            pop_hisp           pop_other         pop_overlap 
              1.014               1.010               1.007               1.011               1.001 
          pop_white total_county_splits   total_muni_splits           total_vap            vap_aian 
              1.009               1.000               1.001               1.008               1.006 
          vap_asian           vap_black            vap_hisp           vap_other           vap_white 
              1.006               1.012               1.006               1.010               1.007 
• R-hat ≤ 1.05: 325
• 1.05 < R-hat ≤ 1.1: 0
• R-hat > 1.1: 0
• Total R-hats: 325
Sampling diagnostics for SMC run 1 of 5 (5,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique 
Split 1    14,203 (71.0%)      0.5%        0.69 12,591 (100%) 
Split 2     7,189 (35.9%)     97.5%        1.04 11,411 ( 90%) 
Split 3    10,435 (52.2%)     68.0%        0.78  9,346 ( 74%) 
Split 4    12,962 (64.8%)     48.6%        0.63 10,341 ( 82%) 
Split 5    14,578 (72.9%)     36.1%        0.54 10,958 ( 87%) 
Split 6    15,669 (78.3%)     27.7%        0.47 11,290 ( 89%) 
Split 7    16,488 (82.4%)     21.7%        0.42 11,358 ( 90%) 
Split 8    17,275 (86.4%)     17.6%        0.38 11,367 ( 90%) 
Split 9    17,825 (89.1%)     14.8%        0.34 11,130 ( 88%) 
Split 10   18,297 (91.5%)     12.6%        0.30 11,117 ( 88%) 
Split 11   18,790 (94.0%)     10.8%        0.26 10,881 ( 86%) 
Split 12   19,358 (96.8%)      9.2%        0.19 10,464 ( 83%) 
Resample   19,358 (96.8%)       NA%          NA 18,532 (147%) 

Sampling diagnostics for Mergesplit Steps of SMC run 1 of 5 (5,000 samples)
          Acc. rate MS Steps per Plan
MS Step 1     36.4%               300
MS Step 2     37.7%               150
MS Step 3     39.2%               150
MS Step 4     40.6%               150
MS Step 5     41.9%               150

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of
the log weights (more than 3 or so), and low numbers of unique plans. R-hat values for summary
statistics should be between 1 and 1.05.
```

```
  n_black_perf    n
1            1 5000
```

```
# A tibble: 13 × 2
   bvap_rank   dem
       <dbl> <dbl>
 1         1 0.082
 2         2 0.826
 3         3 0.964
 4         4 0.993
 5         5 0.992
 6         6 0.948
 7         7 0.987
 8         8 0.999
 9         9 1    
10        10 1    
11        11 1    
12        12 1    
13        13 1    
```

## Checklist

- [x] I have followed the [instructions](https://github.com/alarm-redist/fifty-states/blob/main/CONTRIBUTING.md)
- [x] I have updated the [tracker](https://docs.google.com/spreadsheets/d/1k_tYLoE49W_DCK1tcWbouoYZFI9WD76oayEt5TOmJg4/edit#gid=453387933)
- [x] All `TODO` lines from the template code have been removed
- [x] I have merged in the main branch and then recalculated summary statistics
- [x] I have run `enforce_style()` to format my code
- [x] The documentation copied above is up-to-date 
- [x] There are no data files in this pull request
- [x] None of the file output paths (for the `redist_map` and `redist_plans` objects, and summary statistics) have been edited

@tylersimko

brianlzhou committed:
init

brianlzhou committed:
Update 01_prep_NJ_cd_1990.R

brianlzhou committed:
edits

brianlzhou committed:
todo

tylersimko commented:
@brianlzhou Looks like we're not quite matching the top two minority VAP districts. Have you checked performance in the two? Probably okay, but good to be sure.

brianlzhou committed:
vals

tylersimko commented on analyses/NJ_cd_1990/03_sim_NJ_cd_1990.R:
```diff
@@ -0,0 +1,42 @@
+cli_process_start("Running simulations for {.pkg NJ_cd_1990}")
+
+set.seed(1990)
+plans <- redist_smc(map, nsims = 1e4, runs = 5, counties = county, pop_temper = 0.015)
```
Just pinging @brianlzhou

brianlzhou committed:
Update 03_sim_NJ_cd_1990.R

brianlzhou committed:
Update 01_prep_NJ_cd_1990.R

brianlzhou committed:
Update 01_prep_NJ_cd_1990.R

brianlzhou committed:
Update logit_shift.R

brianlzhou committed:
Update 03_sim_NJ_cd_1990.R

brianlzhou commented:
<img width="3000" height="5400" alt="image" src="https://github.com/user-attachments/assets/4a52fec5-fd4c-4a56-a4b7-f541fb116ebd" />

```
SMC: 5,000 sampled plans of 13 districts on 1,944 units
Plans sampled on Graph Space using the Naive Top K forward kernel.
Forward kernel parameters: `adapt_k_thresh`=0.99
SMC Parameters:
• `weight_type` = optimal
• `seq_alpha` = 1
• `pop_temper` = 0.015
Plan diversity 80% range: 0.35 to 0.59
Largest R-hat values for ordered summary statistics:
    comp_bbox_reock           comp_edge         comp_polsby       county_splits         muni_splits 
              1.024               1.020               1.049               1.002               1.017 
            ndshare                 ndv                 nrv            plan_dev            pop_aian 
              1.032               1.030               1.029               1.025               1.034 
          pop_asian           pop_black            pop_hisp           pop_other         pop_overlap 
              1.031               1.024               1.029               1.031               1.007 
          pop_white total_county_splits   total_muni_splits           total_vap            vap_aian 
              1.025               1.000               1.019               1.030               1.027 
          vap_asian           vap_black            vap_hisp           vap_other           vap_white 
              1.046               1.025               1.018               1.028               1.024 
• R-hat ≤ 1.05: 325
• 1.05 < R-hat ≤ 1.1: 0
• R-hat > 1.1: 0
• Total R-hats: 325
Sampling diagnostics for SMC run 1 of 5 (5,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     6,120 (61.2%)     12.0%        0.66 6,285 ( 99%)      7 
Split 2     6,139 (61.4%)     14.9%        0.49 5,603 ( 89%)      5 
Split 3     5,044 (50.4%)     16.9%        0.53 5,773 ( 91%)      4 
Split 4     4,990 (49.9%)     15.3%        0.52 5,502 ( 87%)      4 
Split 5     4,629 (46.3%)     11.5%        0.55 5,561 ( 88%)      5 
Split 6     5,854 (58.5%)     14.1%        0.60 5,328 ( 84%)      4 
Split 7     5,309 (53.1%)     17.5%        0.55 5,358 ( 85%)      3 
Split 8     4,023 (40.2%)     24.1%        0.62 5,478 ( 87%)      2 
Split 9     6,041 (60.4%)     10.9%        0.55 5,321 ( 84%)      4 
Split 10    5,353 (53.5%)     11.9%        0.56 5,552 ( 88%)      3 
Split 11    4,248 (42.5%)      7.1%        0.67 5,149 ( 81%)      4 
Split 12    9,773 (97.7%)      3.5%        0.16 4,597 ( 73%)      3 
Resample    9,773 (97.7%)       NA%          NA 9,348 (148%)     NA 

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of
the log weights (more than 3 or so), and low numbers of unique plans. R-hat values for summary
statistics should be between 1 and 1.05.
```

brianlzhou committed:
convergence

brianlzhou committed:
edits

brianlzhou committed:
Revert "Update 01_prep_NJ_cd_1990.R"

This reverts commit 233ce9ef98b333b490a8447158fb740df57b4ec7.

brianlzhou committed:
Revert "Update logit_shift.R"

This reverts commit 0eee26a899b52e4c19187c58e966dc09015fd8c5.

brianlzhou committed:
style

christopherkenny commented on analyses/NJ_cd_1990/03_sim_NJ_cd_1990.R:
```diff
@@ -0,0 +1,102 @@
+cli_process_done()
+
+# read in from FASRC
+map <- readRDS(
```
We should drop this type of re-reading code so that it can be run in one go.

christopherkenny commented:
What happened here that we're running 100k total simulations?

christopherkenny commented:
In general, oversampling should be very small with the new algorithm, so I don't quite follow what's going on here.

brianlzhou commented:
I was running into rhat convergence issues with 3000 simulations, and this persisted for 5000 as well, so I bumped it to 20000. I could run an intermediate of 10000 if you'd prefer?
